src/geometry.F90

```diff
@@ -770,34 +649,48 @@ end subroutine cross_lattice
 ! that has a parent cell, also include the surfaces of the edge of the universe.
 !===============================================================================
 
-  subroutine distance_to_boundary(p, dist, surface_crossed, lattice_crossed)
+  subroutine distance_to_boundary(p, dist, surface_crossed, lattice_translation)
 
     type(Particle), intent(inout) :: p
     real(8),        intent(out)   :: dist
     integer,        intent(out)   :: surface_crossed
-    integer,        intent(out)   :: lattice_crossed
-
-    integer :: i            ! index for surface in cell
-    integer :: index_surf   ! index in surfaces array (with sign)
-    real(8) :: x,y,z        ! particle coordinates
-    real(8) :: u,v,w        ! particle directions
-    real(8) :: d            ! evaluated distance
-    real(8) :: x0,y0,z0     ! coefficients for surface
-    real(8) :: r            ! radius for quadratic surfaces
-    real(8) :: tmp          ! dot product of surface normal with direction
-    real(8) :: a,b,c,k      ! quadratic equation coefficients
-    real(8) :: quad         ! discriminant of quadratic equation
-    logical :: on_surface   ! is particle on surface?
+    integer,        intent(out)   :: lattice_translation(3)
+
+    integer :: i                 ! index for surface in cell
+    integer :: index_surf        ! index in surfaces array (with sign)
+    real(8) :: x,y,z             ! particle coordinates
+    real(8) :: beta, gama        ! skewed particle coordiantes
+    real(8) :: u,v,w             ! particle directions
+    real(8) :: beta_dir          ! skewed particle direction
+    real(8) :: gama_dir          ! skewed particle direction
+    real(8) :: edge              ! distance to oncoming edge
+    real(8) :: d                 ! evaluated distance
+    real(8) :: x0,y0,z0          ! coefficients for surface
+    real(8) :: r                 ! radius for quadratic surfaces
+    real(8) :: tmp               ! dot product of surface normal with direction
+    real(8) :: a,b,c,k           ! quadratic equation coefficients
+    real(8) :: quad              ! discriminant of quadratic equation
+    logical :: on_surface        ! is particle on surface?
     type(Cell),       pointer, save :: cl => null()
     type(Surface),    pointer, save :: surf => null()
-    type(Lattice),    pointer, save :: lat => null()
+    class(Lattice),   pointer, save :: lat => null()
     type(LocalCoord), pointer, save :: coord => null()
```

Comment:
This goes back a long way to when threading was first being implemented. I think the explanation is as follows: if a variable is initialized at the time it's declared, it implicitly has the save attribute. By virtue of being saved, it is treated as shared data, and therefore if you want it to be threadprivate you must explicitly declare it as such. Conversely, a threadprivate variable must be declared with the save attribute. We may be able to get rid of all the threadprivate and save declarations by not initializing pointers in procedures.